unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp

```diff
@@ -0,0 +1,323 @@
+          typet &sub_type=generic_type_var.subtype();
+          REQUIRE(sub_type.id()==ID_symbol);
+          symbol_typet &bound_type=to_symbol_type(sub_type);
+          REQUIRE(bound_type.get_identifier()=="java::java.lang.Number");
+          REQUIRE(
+            to_symbol_type(
+              java_generics_class_type_bound(0, java_generics_class_type))
+            .get_identifier()=="java::java.lang.Number");
+        }
+
+        const struct_union_typet::componentt &elem=
+          java_generics_class_type.get_component("elem");
+        const typet &elem_type=java_class_type.component_type("elem");
+
+        REQUIRE(is_java_generic_parameter(elem_type));
+      }
+    }
+  }
+
+  GIVEN("Some class files with generic type variable")
+  {
+    WHEN("Parsing a class with bounded type variable")
+    {
+      REQUIRE(new_symbol_table.has_symbol("java::generics"));
+
+      THEN("The generic fields should be annotated with concrete types")
+      {
+        const symbolt &class_symbol=new_symbol_table.lookup("java::generics");
+        const typet &symbol_type=class_symbol.type;
+
+        REQUIRE(symbol_type.id()==ID_struct);
+        class_typet class_type=to_class_type(symbol_type);
+        REQUIRE(class_type.is_class());
+        java_class_typet java_class_type=to_java_class_type(class_type);
+        REQUIRE(!is_java_generics_class_type(java_class_type));
+
+        const struct_union_typet::componentt &belem=
+          java_class_type.get_component("belem");
+        const typet &belem_type=java_class_type.component_type("belem");
+
+        REQUIRE(belem_type!=nil_typet());
+        REQUIRE(is_java_generic_type(belem_type));
+        THEN("Field has instantiated type variable")
+        {
+          const java_generic_typet &container=
+            to_java_generic_type(belem_type);
+
+          const std::vector<java_generic_parametert> &generic_types=
+            container.generic_type_variables();
+          REQUIRE(generic_types.size()==1);
+
+          const typet& inst_type=java_generic_get_inst_type(0, container);
+
+          REQUIRE(inst_type.id()==ID_pointer);
+          const typet &inst_type_symbol=inst_type.subtype();
+          REQUIRE(inst_type_symbol.id()==ID_symbol);
+          REQUIRE(
+            to_symbol_type(inst_type_symbol).get_identifier()==
+            "java::java.lang.Integer");
+        }
+      }
+    }
+  }
+
+  GIVEN("Some class files with Generics")
+  {
+    WHEN("Methods with generic signatures")
+    {
+      REQUIRE(
+        new_symbol_table
+          .has_symbol("java::generics$bound_element.f:()Ljava/lang/Number;"));
+
+      THEN("The method should have generic return type")
+      {
+        const symbolt &method_symbol=
+          new_symbol_table
+            .lookup("java::generics$bound_element.f:()Ljava/lang/Number;");
+        const typet &symbol_type=method_symbol.type;
+
+        REQUIRE(symbol_type.id()==ID_code);
+
+        const code_typet &code=to_code_type(symbol_type);
```

Comment:
This test case doesn't actually test what is stated in the 'THEN' clause - and I think this functionality has broken. I've just rebased my work on top of this PR, and one of my unit tests is failing because of this. Previously, you could call `is_java_generic_parameter(code.return_type())` and get the answer `true` for this function. I'm not finding this is returning `false` instead, and doing 'method_symbol.show(std::cout)` is showing that the return type for this method is not marked as being generic in any way.

Reply:
Thanks for this, this has also been failing the CI tests. The reason is that the return type of the lookup method for symbol table has changed (commit #1444 ) - now it's a reference wrapper around symbolt instead of symbolt.

Reply:
Yes, that is part of the failure (it's broken my work too :-) ), but my broader point is that this test case doesn't *actually* test what it claims to test - it claims to test that it parses methods with a generic return type, but it only does that partially - it checks that it can parse the method, and that the symbol for the method exists, and is an ID_code object - but it *doesn't* check that the return type of that `code_typet` actually records that it is generic. So, for instance, we could be parsing the the method and then throwing away all the generic information and this test as it is currently written would still be passing. This appears to have happened at some point, because when my code was based off @thk123 branch, the `is_java_generic_parameter(code.return_type())` was correctly returning true for this case, whereas with this branch I'm seeing that the generic information is no longer stored for the return type.

Reply:
Yes - this is the mess with descriptors of methods being used. As discussed I think before we do [TG-865](https://diffblue.atlassian.net/browse/TG-865) it won't work. The THEN should be updated in a similar way to the TODO directly below.